Switch error tracking to Sentry

rakyi opened the pull request:
* Remove all usage of Bugsnag
* Add Sentry to admin client bundle and all server side processes
  * Move the Sentry initialization to be as early as possible to maximize the chance an error will be captured
  * Include node profiling integration for server side
  * Create a new `admin` project for the admin client and server processes
* Tell Sentry about:
  * `release` is the `COMMIT_SHA` of the currently running code - helps with triage and works with GitHub integration for stuff like comments on PRs and other features
  * `environment` is good for filtering issues, we are gonna distinguish between prod and staging from now on. Running Sentry on staging should help us catch issues before they get to prod.
* Add GitHub action for creating Sentry releases
  * Telling Sentry about a release before it gets the first event from that version makes some features nicer to use (I think e.g. "Mark as resolved in the next release")
* Add Sentry Vite plugin - used to upload source maps as part of the production build. This is the recommended way to do it and AFAIK the easiest to get client source maps to work also in staging.
* Rework manual capturing of errors:
  * Rename `logErrorAndMaybeSendToBugsnag` to `logErrorAndMaybeCaptureInSentry`
  * Where it makes sense, throw an error instead of calling this function and let the Sentry error handler capture it
  * Remove two instances of logging gdoc validation errors and warnings during baking, since we get more than 1k of them in one bake, which makes them useless and consumes a lot of our quota. We should add it back only if we making it smarter or fix most of the errors first.
  * Remove unnecessary `warn` wrapper for `console.warn`
  * Change `JsonError` to `Error` where it isn't necessary
* Increase Bundlewatch limit substantially - Sentry dependencies were previously not included in the build at all, giving us a false sense of the bundle being smaller than it was

You can see various testing issues and new issues from baking etc. from this branch as well as current ones from production in [Sentry](https://our-world-in-data.sentry.io).

Resolves #4204

Depends on https://github.com/owid/ops/pull/263

rakyi commented:
Bundle watch originally failed here with a large bundle size increase. We are in fact making the bundle smaller from 2.91 MB to 2.71 MB, but Bundlewatch didn't pick it up, because it seems that rollup is smart enough to remove code that won't run at import time when it's guarded by a falsy condition, i.e. the SENTRY_DSN isn't set in CI, so the code that was using Sentry didn't run, so rollup didn't include in the bundle previously.

We have now removed the condition and initialize Sentry always (which is fine to do and fails gracefully without DSN). So rollup now doesn't remove the code, and Bundlewatch thus detects the bundle increase.

ikesau commented on site/instrument.ts:
```diff
@@ -0,0 +1,35 @@
+import { COMMIT_SHA, ENV, SENTRY_DSN } from "../settings/clientSettings.js"
+import { getPreferenceValue, PreferenceType } from "./cookiePreferences.js"
+
+const analyticsConsent = getPreferenceValue(PreferenceType.Analytics)
```
Same as before, but this means we'll never be able to log sessions from new users on their first load of a page?

i.e.
1. Load homepage
2. Sentry initializes without session tracking
3. Click "Accept cookies"

I guess it wouldn't make sense to begin recording a session midway (though it would be kind of funny to see them "turning the camera on" by clicking the "Accept cookies" button.

Oh well. Maybe check with Bobbie if he thinks these new user sessions would be especially useful, but my sense is that this is fine.

ikesau commented on site/instrument.ts:
```diff
@@ -0,0 +1,35 @@
+    }
+} else {
+    sentryOpts = {
+        replaysSessionSampleRate: 0,
```
Do these need to be set at all if `Sentry.replayIntegration` isn't added?

rakyi replied:
Not really, but I think Bobbie wanted to be more explicit so e.g. you don't have to look up the docs to know what's the default and I think it's fine to keep it as is.

rakyi replied:
On the other hand, I think you are right, it makes less sense when the `replayIntegration` ins't added. Changing.

ikesau reviewed (commented):
Really nice to have someone go through this all and make it sane.

There's one other random mention of Bugsnag in the codebase (a comment explaining why some object isn't desctructured) which I think we can also remove.

I think doing some tire kicking over a video call would be a good way to test this. Wanna book me for 15 minutes? 🙂

rakyi commented on adminSiteClient/instrument.ts:
```diff
@@ -0,0 +1,7 @@
+import * as Sentry from "@sentry/react"
+
+Sentry.init({
+    dsn: process.env.SENTRY_ADMIN_DSN,
```
The idea is that you want to run as little code as possible before initializing Sentry, and since there can be an error in `clientSettings`, you want it to be captured by Sentry. But I noticed some issues with relying on `process.env` on the client side (prod vs dev build) and it seems we need `clientSettings` after all for it to work reliably.

Changed.

rakyi commented on adminSiteServer/authentication.ts:
```diff
@@ -100,7 +100,7 @@ export async function authCloudflareSSOMiddleware(
         httpOnly: true,
         sameSite: "lax",
-        secure: ENV === "production",
+        secure: ENV !== "development",
```
I don't think it's worth maintaining such a list separately from the code. If I saw a list like that, I wouldn't trust it and would search through the code anyway because of the high likelihood of it being out of sync.

rakyi commented on .github/workflows/sentry.yml:
```diff
@@ -0,0 +1,21 @@
+                  SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
+                  SENTRY_ORG: ${{ secrets.SENTRY_ORG }}
+              with:
+                  environment: ${{ github.ref_name == 'master' && 'production' || 'staging' }}
```
Yeah, it's not very nice. AFAIK, it's not really YAML, but rather a GitHub actions specific syntax.

rakyi commented on baker/SiteBaker.tsx:
```diff
@@ -208,9 +211,6 @@ export class SiteBaker {
         this.explorerAdminServer = new ExplorerAdminServer(GIT_CMS_DIR)
-        // As of 10-10-2024, baking sends too many events to Bugsnag which causes us to hit the rate limit.
-        // We'll have to find a way to be more selective about how we log baking errors/warnings.
-        // (Only sending errors probably isn't sufficient, because there are some warnings that we want to know about.)
```
Thanks, this one somehow slipped through.

rakyi commented on packages/@ourworldindata/grapher/src/core/Grapher.tsx:
```diff
@@ -1,5 +1,6 @@
 import React from "react"
 import ReactDOMServer from "react-dom/server.js"
+import * as Sentry from "@sentry/react"
```
Exactly.

rakyi commented on adminSiteClient/instrument.ts:
```diff
@@ -0,0 +1,7 @@
+import * as Sentry from "@sentry/react"
```
I'd like to keep the name, because:

* `instrument` is what is used in Sentry docs, so if you read the docs and then search for such file, you'll easily find all three of them. Scope is in the module path, i.e. the directory name, we don't usually repeat that in nested file names.
* `instrument` is named by what it does, so if we ever want to add instrumentation of something else, we can.
* We don't have to change the name, when we change providers. We also don't call files/folders as specific as `mysql` when we don't need to, but rather `db`.

ikesau commented:
# Manual testing notes.

## 1.

![image](https://github.com/user-attachments/assets/fc3e9f9b-b003-40a3-a9fc-8c62e1fe0d49)

This is from trying to add a circular redirect (which is inferrable through the stack trace) but I'm wondering if this is the top-level message given for all the JsonErrors?


## 2.

![image](https://github.com/user-attachments/assets/273a7986-11e2-4231-8c36-0bb284fe9f65)

OOC, what's up with all these different `ADMIN-X` instances? Do they represent different staging instances? Is there a way we can name them more clearly?


## 3.
[Here's an error from me adding a malformed gdoc](https://our-world-in-data.sentry.io/issues/18257808/?project=4508607592464464&query=release%3A1d2827543711198301ea3c692ee1ead21af59e84%20error.unhandled%3Atrue&referrer=issue-stream&sort=freq&stream_index=1). In Bugsnag, I was able to look at the headers and see the user's email in them, but it looks like they're not here in Sentry? Not essential, but it would be nice to have as I often used this information to quickly follow up with authors who might be stuck on something. 

## 4.
I wonder if we should add some more DSNs - the fact that errors/crashes from baking and indexing are both under `admin` is possible to understand, but if there's not much cost 2 having 2 additional buckets for those issues then maybe that would make things clearer?

rakyi commented:
Thanks a lot for the manual testing!

1. Good catch. The handling of API errors is pretty weird in admin. I switched the code to throw `JsonError` instead of a plain object, which [seems to](https://our-world-in-data.sentry.io/issues/20765861/events/e140fc95557642de8885309d331e1d2d/) have fixed the problem.
2. Labels like `ADMIN-X` are identifiers for Sentry issues unique across an org so you can easily reference them in various places like comments, commit messages, etc.
3. Good point, I forgot to check whether users are associated with requests! I added the code to set a user in Sentry. See [here](https://our-world-in-data.sentry.io/issues/20711864/events/55a9854089b34a4eb7778477efa216b4/) for how it looks with a user.
4. We can have basically as many projects as we want. It's all about finding what works best for a given org/team. In a past job, I was frustrated with having too many of them and IME it worked better with fewer. I'd say let's see if two work well for us and adjust later if necessary.

ikesau reviewed (approved):
Nice!

Let's get this in and ready to demo for knowledge sharing. 🔥
